10-Merge/readme.md

```diff
@@ -0,0 +1,202 @@
+git-merge
+=========
+
+> Join two or more development histories together
+
+We will start with warning from documentation:
+
+> Warning: Running git merge with non-trivial uncommitted changes is discouraged: while possible, it may leave you in a state that is hard to back out of in the case of a conflict.
+
+Let's all start with cloning testing repository to check merge on:
+
+	$ git clone git@github.com:DevTrainings/test_merge_conflict.git
+	$ git checkout bar && git checkout cherry-pick && git checkout foo && git checkout master
+
+This repository looks like this (run it yourself to have it in colors ;) ):
+
+	$ git log --all --graph --oneline --decorate
+	* c5d665f (HEAD -> master, origin/master, origin/HEAD) fixed readme a bit
+	| * 4909911 (origin/cherry-pick, cherry-pick) merge conflict in file
+	| * e7f0e30 added file for cherry_pick
+	|/  
+	* 286a273 added readme
+	| * c5c3da1 (origin/foo, foo) file2 for demonstration of cherry-pick
+	| * c361f2e foo
+	|/  
+	| * bfb5647 (origin/bar, bar) bar
+	|/  
+	* 25f7328 file
+
+We will ignore `cherry-pick` branch for now, we will get to it later. We have file `file` in the root directory
+
+	$ ls -l .
+	total 8
+	-rw-r--r-- 1 paladin paladin  10 Apr 11 10:38 file
+	-rw-r--r-- 1 paladin paladin 154 Apr 11 11:08 readme.md
+	$ cat file
+	.
+	.
+	.
+	.
+	.
+
+The file `file` was changed in both branches `bar` and `foo` and we want to get those changes back into the master. We will start with `bar`
+
+	$ git merge bar
+	Merge made by the 'recursive' strategy.
+	 file | 4 ++--
+	 1 file changed, 2 insertions(+), 2 deletions(-)
+
+> Running git merge with non-trivial uncommitted changes is discouraged: while possible, it may leave you in a state that is hard to back out of in the case of a conflict.
+
+This command produces no conflicts, since no changes we made to `file` after branching into `bar`. So after this merge, we can see that repository looks like this:
+
+	$ git log --all --graph --oneline --decorate
+	*   05dbdf3 (HEAD -> master) Merge branch 'bar'
+	|\  
+	| * bfb5647 (origin/bar, bar) bar
+	* | c5d665f (origin/master, origin/HEAD) fixed readme a bit
+	| | * 4909911 (origin/cherry-pick, cherry-pick) merge conflict in file
+	| | * e7f0e30 added file for cherry_pick
+	| |/  
+	|/|   
+	* | 286a273 added readme
+	|/  
+	| * c5c3da1 (origin/foo, foo) file2 for demonstration of cherry-pick
+	| * c361f2e foo
+	|/  
+	* 25f7328 file
+
+	$ cat file
+	.
+	bar
+	.
+	.
+	bar
+
+We can see that branch `bar` was merged into `master`. Now it's time to do the same with `foo`.
+
+	$ git merge foo
+	Auto-merging file
+	CONFLICT (content): Merge conflict in file
+	Automatic merge failed; fix conflicts and then commit the result.
+
+We have a conflict! We need to somehow resolve it.
+
+> You can always abort merge with `git merge --abort`
+
+Text editor way
+---------------
+
+In case of conflict, git puts changes into a file like this:
```

Comment:
little bit :) but still useful when you sometime need to merge something on server over ssh or something